### PR TITLE
[WIP] BungeeCord support

### PR DESCRIPTION
**Pull request by MrIvanPlays**
This has awfully lot of stuff to do. Thoughts wanted, especially on the way a EconomyProvider would be obtained because BungeeCord does not have a service provider.

### CODE REVIEW

**Comment by lokka30** [comment]
> This has awfully lot of stuff to do. Thoughts wanted, especially on the way a EconomyProvider would be obtained because BungeeCord does not have a service provider.

I hope the other platforms have a service provider.
I think the only solutions are
* make the service registration in Treasury's bungee core (not the API of course)
* make the service registration a separate library for bungee and whichever else requires it

File: platform/bungee/api/src/main/java/me/lokka30/treasury/api/economy/bungee/misc/BungeeCordEventCaller.java [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+
+                callback.done(isCancelled(event, patEvent), null);
+            };
+
+            ProxyServer
+                    .getInstance()
+                    .getPluginManager()
+                    .callEvent(new NonPlayerAccountTransactionEvent(account,
+                            transaction,
+                            pateCallback
```
Comment (lokka30) [inline platform/bungee/api/src/main/java/me/lokka30/treasury/api/economy/bungee/misc/BungeeCordEventCaller.java:113]: shouldn't this be `npate`

File: platform/bungee/api/src/main/java/me/lokka30/treasury/api/economy/bungee/misc/BungeeCordEventCaller.java [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+
+        ProxyServer.getInstance().getPluginManager().callEvent(new AccountTransactionEvent(account,
+                transaction,
+                ateCallback
+        ));
+    }
+
+    private static boolean isCancelled(Cancellable event, Cancellable event1) {
+        boolean cancelled;
+        if (event.isCancelled() && event1.isCancelled()) {
```
Comment (lokka30) [inline platform/bungee/api/src/main/java/me/lokka30/treasury/api/economy/bungee/misc/BungeeCordEventCaller.java:125]: probably better just to make this a big return statement

Comment (lokka30) [reply platform/bungee/api/src/main/java/me/lokka30/treasury/api/economy/bungee/misc/BungeeCordEventCaller.java:125]: also no need for the `cancelled` variable, directly return in the statements

File: platform/bungee/plugin/pom.xml [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+            <version>${bungeecord.version}</version>
+            <scope>provided</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.jetbrains</groupId>
+            <artifactId>annotations</artifactId>
+            <version>${jetbrains.annotations.version}</version>
+            <scope>provided</scope>
+        </dependency>
+        <!-- TODO: bStats?
```
Comment (lokka30) [inline platform/bungee/plugin/pom.xml:114]: how come did we comment it out?

**Review by lokka30 — COMMENTED** [review]
Thank you very much for working on this!

File: platform/bungee/plugin/pom.xml (hunk 3 above)
Comment (MrIvanPlays) [reply platform/bungee/plugin/pom.xml:114]: I commented it out with a todo because bungee metrics require a new page on bstats.

**Comment by MrIvanPlays** [comment]
> > This has awfully lot of stuff to do. Thoughts wanted, especially on the way a EconomyProvider would be obtained because BungeeCord does not have a service provider.
> 
> 
> 
> I hope the other platforms have a service provider.

Thing is velocity also doesn't have one. The difference is that velocity is willing to get one whilst md_5 is strongly against adding such in bungee.

**Comment by lokka30** [comment]
> > > This has awfully lot of stuff to do. Thoughts wanted, especially on the way a EconomyProvider would be obtained because BungeeCord does not have a service provider.
> > 
> > 
> > I hope the other platforms have a service provider.
> 
> Thing is velocity also doesn't have one. The difference is that velocity is willing to get one whilst md_5 is strongly against adding such in bungee.

Should we add a service registration methods into the Treasury Bungee Core?

**Comment by MrIvanPlays** [comment]
Thats what I did in the Bungee API but I don't think it is the best implementation. I'd like @Jikoo's opinion on this

**Comment by MrNemo64** [comment]
Since Bungee doesn't have the service provider, with your current implementation, we would need a diferent provider class for every api. 
In case we decide to go with the option of creating our own service provider (which I personally can't think of a better way), I think it should work to provide any API, not just the EconomyProvider

**Comment by MrIvanPlays** [comment]
what im also thinking of is our own event handler. this will eliminate the need for per platform apis which would basically mean that treasury api could be implemented also in non minecraft related programs ( if anyone would use it for such )

**Comment by MrIvanPlays** [comment]
looks like im trying to do stuff too soon. closing this and im gonna open a discussion tab about the problems.

**Pull request closed without merging**